packages/ra-core/src/controller/input/useReferenceInputController.ts

```diff
@@ -77,61 +60,195 @@ interface Option {
-        resource,
-        source,
-        id: input.value,
-    });
+    // selection logic
+    const {
+        selectedIds,
+        onSelect,
+        onToggleItem,
+        onUnselectItems,
+    } = useSelectionState();
+
+    // fetch possible values
+    const {
+        ids: possibleValuesIds,
+        data: possibleValuesData,
+        total: possibleValuesTotal,
+        loaded: possibleValuesLoaded,
+        loading: possibleValuesLoading,
+        error: possibleValuesError,
+    } = useGetList(reference, pagination, sort, filterValues);
 
-    const { referenceRecord } = useReference({
+    // fetch current value
+    const {
+        referenceRecord,
+        error: referenceError,
+        loading: referenceLoading,
+        loaded: referenceLoaded,
+    } = useReference({
         id: input.value,
         reference,
     });
 
+    // add current value to possible sources
+    let finalIds: Identifier[],
+        finalData: RecordMap<Record>,
+        finalTotal: number;
+    if (!referenceRecord || possibleValuesIds.includes(input.value)) {
+        finalIds = possibleValuesIds;
+        finalData = possibleValuesData;
+        finalTotal = possibleValuesTotal;
+    } else {
+        finalIds = [input.value, ...possibleValuesIds];
+        finalData = { [input.value]: referenceRecord, ...possibleValuesData };
+        finalTotal += 1;
+    }
+
+    // overall status
     const dataStatus = getDataStatus({
         input,
-        matchingReferences,
+        matchingReferences: Object.keys(finalData).map(id => finalData[id]),
         referenceRecord,
         translate,
     });
 
     return {
-        choices: dataStatus.choices,
+        // should match the ListContext shape
+        possibleValues: {
+            basePath,
+            data: finalData,
+            ids: finalIds,
+            total: finalTotal,
+            error: possibleValuesError,
+            loaded: possibleValuesLoaded,
+            loading: possibleValuesLoading,
+            hasCreate: false,
+            page,
+            setPage,
+            perPage,
+            setPerPage,
+            currentSort: sort,
+            setSort,
+            filterValues,
+            displayedFilters,
+            setFilters: setFilter,
+            showFilter,
+            hideFilter,
+            selectedIds,
+            onSelect,
+            onToggleItem,
+            onUnselectItems,
+            resource,
+        },
+        referenceRecord: {
+            data: referenceRecord,
+            loaded: referenceLoaded,
+            loading: referenceLoading,
+            error: referenceError,
+        },
+        dataStatus: {
+            error: dataStatus.error,
+            loading: dataStatus.waiting,
+            warning: dataStatus.warning,
+        },
+        choices: Object.keys(finalData).map(id => finalData[id]),
+        // kept for backwards compatibility
+        // @deprecated to be removed in 4.0
         error: dataStatus.error,
```

Comment:
Only the `error` is deprecated or it's also the following parameters?

Reply:
it's the following parameters, but we'll need the RecordContext to truly deprecate them